rust-bot/src/main.rs

```diff
@@ -1,41 +1,108 @@
-                println!("Error sending message: {:?}", why);
-            }
-        }
-    }
+#[group]
+#[checks(Mod)]
+#[commands(create_cohort)]
+struct Mod;
 
-    // Set a handler to be called on the `ready` event. This is called when a
-    // shard is booted, and a READY payload is sent by Discord. This payload
-    // contains data like the current user's guild Ids, current user data,
-    // private channels, and more.
-    //
-    // In this case, just print what the current user's username is.
-    fn ready(&self, _: Context, ready: Ready) {
-        println!("{} is connected!", ready.user.name);
+#[command]
+fn ping(ctx: &mut Context, msg: &Message) -> CommandResult {
+    if let Err(why) = msg.channel_id.say(&ctx.http, "Pong!") {
+        println!("Error sending message: {}", why);
     }
+    return Ok(());
+}
+
+// This command provisions out a channel with permissions
+// to read/write for users with the corresponding role. It
+// will also make the channel read-only for other users.
+#[command]
+fn create_cohort(ctx: &mut Context, msg: &Message, mut args: Args) -> CommandResult {
+    msg.channel_id
+        .say(&ctx.http, "Spinning up Adventure Club...")?;
+
+    let (cohort_name, channel_name) = gen_names(args.single::<String>().unwrap());
+    let role = create_role(ctx, msg, &cohort_name);
+    let channel = create_channel(ctx, msg, &channel_name, role);
+
+    match channel {
+        Ok(channel) => {
+            let reply_msg = format!(
+                "Successfully created {}! Feel free to add users.",
+                channel.name
+            );
+            msg.reply(&ctx.http, reply_msg).unwrap();
+        }
+        Err(_) => {
+            msg.reply(&ctx.http, "Failed to create cohort").unwrap();
+        }
+    };
+
+    Ok(())
+}
+
+fn create_role(ctx: &mut Context, msg: &Message, role_name: &str) -> Role {
+    let guild = msg.guild(&ctx.cache).unwrap();
+    let role = match guild.read().role_by_name(role_name) {
+        Some(role) => {
+            let content = format!("{} Role already exists!", role.name);
+            if let Err(error) = msg.channel_id.say(&ctx.http, content) {
+                println!("{:?}", error);
+            };
+            role.clone()
+        }
+        None => guild
+            .read()
+            .create_role(&ctx, |r| r.name(role_name))
+            .unwrap(),
+    };
+    return role;
+}
+
+// Permissions for chatting happen here
+fn create_channel(
+    ctx: &mut Context,
+    msg: &Message,
+    channel_name: &str,
+    role: Role,
+) -> Result<GuildChannel, Error> {
+    let role_id = role.id;
+    let everyone_id = get_everyone_role(ctx, msg).unwrap().id;
+    let permission_set = mute_users_without_role_permset(role_id, everyone_id);
+
+    let guild = msg.guild(&ctx.cache).unwrap();
+    let new_channel = guild.read().create_channel(&ctx.http, |c| {
+        c.name(channel_name)
+            .kind(ChannelType::Text)
+            .permissions(permission_set)
```

Comment:
Add mods to permissions set, with the ability to do anything that is denied to `@everyone`